detox/src/devices/Device.js

```diff
@@ -5,119 +5,95 @@ const { traceCall } = require('../utils/trace');
+      this._cachedAppLaunchArgs = new LaunchArgs(this._getCurrentApp().launchArgs);
     }
 
-    const baseLaunchArgs = {
-      ...this._launchArgs.get(),
-      ...params.launchArgs,
-    };
+    return this._cachedAppLaunchArgs;
+  }
 
-    if (params.url) {
-      baseLaunchArgs['detoxURLOverride'] = params.url;
-      if (params.sourceApp) {
-        baseLaunchArgs['detoxSourceAppOverride'] = params.sourceApp;
-      }
-    } else if (params.userNotification) {
-      this._createPayloadFileAndUpdatesParamsObject('userNotification', 'detoxUserNotificationDataURL', params, baseLaunchArgs);
-    } else if (params.userActivity) {
-      this._createPayloadFileAndUpdatesParamsObject('userActivity', 'detoxUserActivityDataURL', params, baseLaunchArgs);
-    }
+  async prepare() {
+    await this.deviceDriver.prepare();
 
-    if (params.permissions) {
-      await this.deviceDriver.setPermissions(deviceId, bundleId, params.permissions);
-    }
+    this._deviceId = await traceCall('acquireDevice', () =>
+      this.deviceDriver.acquireFreeDevice(this._deviceConfig.device));
 
-    if (params.disableTouchIndicators) {
-      baseLaunchArgs['detoxDisableTouchIndicators'] = true;
+    const appAliases = Object.keys(this._appsConfig);
+    if (appAliases.length === 1) {
+      await this.selectApp(appAliases[0]);
     }
+  }
 
-    if (this._isAppInBackground(params, bundleId)) {
-      if (hasPayload) {
-        await this.deviceDriver.deliverPayload({...params, delayPayload: true});
-      }
+  async selectApp(name) {
+    if (name === undefined) {
+      throw this._errorBuilder.cantSelectEmptyApp();
     }
 
-    let processId;
-    if (this._behaviorConfig.launchApp === 'manual') {
-      processId = await this.deviceDriver.waitForAppLaunch(deviceId, bundleId, this._prepareLaunchArgs(baseLaunchArgs), params.languageAndLocale);
-    } else {
-      processId = await this.deviceDriver.launchApp(deviceId, bundleId, this._prepareLaunchArgs(baseLaunchArgs), params.languageAndLocale);
-      await this.deviceDriver.waitUntilReady();
-      await this.deviceDriver.waitForActive();
+    if (name === null) { // Internal use to unselect the app
+      this._currentApp = null;
+      return;
     }
-    this._processes[bundleId] = processId;
 
-    await this._emitter.emit('appReady', {
-      deviceId,
-      bundleId,
-      pid: processId,
-    });
-
-    if(params.detoxUserNotificationDataURL) {
-      await this.deviceDriver.cleanupRandomDirectory(params.detoxUserNotificationDataURL);
-    }
-
-    if(params.detoxUserActivityDataURL) {
-      await this.deviceDriver.cleanupRandomDirectory(params.detoxUserActivityDataURL);
+    const appConfig = this._appsConfig[name];
+    if (!appConfig) {
+      throw this._errorBuilder.cantFindApp(name);
     }
-  }
-
-  get appLaunchArgs() {
-    return this._launchArgs;
-  }
 
-  get id() {
-    return this._deviceId;
+    this._currentApp = appConfig;
+    this._cachedAppLaunchArgs = null;
```

Comment:
(Rubber-duck'ing) So this will work but would omit any previous changes made in the context of the selected app. For example:
```js
device.selectApp('app1');
device.appLaunchArgs.modify(...);
device.selectApp('app2'); // changes to app1's args are gone...
device.selectApp('app1'); // ... so here we start from scratch
```

That's ok under the assumption of that `selectApp()` is meant to be called once per test. Was that the (implicit) assumption here?

Reply:
> but... but why?
> and also... why? 😢

Exactly to avoid this scenario you demonstrated.

Reply:
Anyway, we've discussed the motivation in 1:1 mode — although, good to replace that mutation-reliant approach with some more sturdy.